kong/db/dao/certificates.lua

```diff
@@ -0,0 +1,211 @@
+local singletons = require "kong.singletons"
+local cjson      = require "cjson"
+local utils      = require "kong.tools.utils"
+
+-- Get an array of SNI names from either a string (split+sort),
+-- an array(sort) or ngx.null(return {})
+-- Returns an error if the list has duplicates
+-- Returns nil if input is falsy.
+local function parse_name_list(input, errors)
+  local name_list
+  if type(input) == "string" then
+    name_list = utils.split(input, ",")
+
+  elseif type(input) == "table" then
+    name_list = utils.shallow_copy(input)
+
+  elseif input == ngx.null then
+    name_list = {}
+  end
+
+  if not name_list then
+    return nil
+  end
+
+  local found = {}
+  for _, name in ipairs(name_list) do
+    if found[name] then
+      local msg   = "duplicate sni in request: " .. name
+      local err_t = errors:invalid_input(msg)
+      return nil, tostring(err_t), err_t
+    end
+    found[name] = true
+  end
+
+  table.sort(name_list)
+  return setmetatable(name_list, cjson.empty_array_mt)
+end
+
+
+local _Certificates = {}
+
+-- Creates a certificate
+-- If the provided cert has a field called "snis" it will be used to generate server
+-- names associated to the cert, after being parsed by parse_name_list.
+-- Returns a certificate with the snis sorted alphabetically.
+function _Certificates:insert_with_name_list(cert)
+  local db = singletons.db
```

Comment:
This is one of my biggest concerns with this PR. We need to find a way to avoid this at **all costs**.

Reply:
I think I can change `kong.db.dao.init` so that `db` is an attribute of `self` here, so we can do `self.db.snis:whatever` instead of `singletons.db.snis:whatever`. Would that be ok?